src/codegen/encoding/scale_encoding.rs

```diff
@@ -108,12 +242,39 @@ impl AbiEncoding for ScaleEncoding {
         returns
     }
 
-    fn cache_storage_loaded(&mut self, _arg_no: usize, _expr: Expression) {
-        unreachable!("This function is not needed for Scale encoding");
+    fn storage_cache_insert(&mut self, arg_no: usize, expr: Expression) {
+        self.storage_cache.insert(arg_no, expr);
+    }
+
+    fn storage_cache_remove(&mut self, arg_no: usize) -> Option<Expression> {
+        self.storage_cache.remove(&arg_no)
     }
 
-    fn get_encoding_size(&self, _expr: &Expression, _ty: &Type, _ns: &Namespace) -> Expression {
-        unreachable!("This function is not needed for Scale encoding");
+    fn get_encoding_size(
+        &self,
+        expr: &Expression,
+        ty: &Type,
+        _ns: &Namespace,
+        vartab: &mut Vartable,
+        cfg: &mut ControlFlowGraph,
+    ) -> Expression {
+        match ty {
+            Type::String | Type::DynamicBytes => {
+                // When encoding a variable length array, the total size is "compact encoded array length + N elements"
+                let length = Expression::Builtin(
+                    Codegen,
+                    vec![Uint(32)],
+                    Builtin::ArrayLength,
+                    vec![expr.clone()],
+                );
+                if self.is_packed() {
+                    length
+                } else {
+                    increment_by(encode_compact(&length, None, None, vartab, cfg), length)
```

Comment:
It is quite expensive to calculate the length of the compact encoding. 

It is cheaper to just use 4 here, and then adjust the length of the vector after encoding.

Reply:
I like the idea, but I need to think about whether this will work for all cases

Reply:
Just checked. There are many instances where we call into the `abi_encode` codegen function, but the returned length is ignored. Which means random tests start to blow up if I introduce your proposed optimization.

This in turn means even more changes for this PR that are not directly related to the problem I'm trying to solve, so I'm quite averse implementing this optimization now. However, I think this is a good discovery and we should change every single instance to use the returned length instead, so we have a single source of truth. Then we should also be able to introduce this as an optimization.

Reply:
Of course, it was just a suggestion.

Reply:
It might be worth leaving a comment in the code, so a potential optimization is not forgotten.

Reply:
Good idea